<a href="https://colab.research.google.com/github/askot19/Analise_de_dados/blob/main/Pokemon/Pokemon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Data Analyst**

*by [Paulo Henrique de Melo Ferreira](https://www.linkedin.com/in/paulo-henrique-ferreira-20151374/)*

---

<img alt = "pokemon logo" width = "35%" src = "https://logodownload.org/wp-content/uploads/2017/08/pokemon-logo-8.png" align = "center">


# Análise de Dados - Pokémon Competitivo
---
# Introdução

Pokémon é uma franquia japonesa que vem sido lançada desde 1996. Dentre os produtos lançados temos jogos de vídeo game, cardgame, animações de TV, longa metragem, séries de TV, spin offs, brinquedos, etc.

Acerca dos jogos de vídeo game, sabemos que é uma indústria que move grandes quantias, seja na venda de produtos e consoles ou mesmo em campeonatos de jogos digitais. A critério de exemplo, é sabido que apenas o aplicativo Pokémon GO foi capaz de arrecadar cerca de 3,3 bilhões de reais no ano de 2021.

Sendo assim, os grandes campeonatos de disputa digital também são responsáveis por montantes de valores astronômicos, naquilo que englobam locação de espaços para realização de eventos, organização, contratação e treinamento de equipe especializada, estadia de competidores, compra dos jogos e consoles da fanquia, ingressos e inscrições nas competições, premiações, etc.

Sendo assim, uma das maneiras em que podemos aplicar a análise de dados seria na seleção de um time capaz de competir e levar uma possível premiação em uma competição de combate digital.


# Objetivos

Este relatório visa selecionar um time de seis Pokémon para se enfrentar campeonatos do cenário competitivos. Os critérios a serem levados em consideração são os atributos dos Pokémon e algumas limitações estabelecidas pelos concurso:

1. Pokémon não lendários, uma vez que estes são banidos de competições;
2. Pokémon com maior ataque para fazer um fronte específico;
3. Pokémon com a maior defesa para se obter o melhor tanque;
4. Pokémon com o maior ataque especial, para se obter o melhor dano não elemental;
5. Pokémon com maior defesa especial, para se obter o mais resistente;
6. Pokémon com a maior velocidade, para se obter um coringa estratégico;
7. Os tipos não podem se repetir mais de duas vezes.

Ao final faremos uma verificação se o time selecionado é composto por Pokémon contidos no META, que seria a seleção daqueles melhores para competições. Como os times de campeonato são, geralmente, compostos de 6 Pokémon, o último será adotado como Pokémon coringa, um Pokémon não elemental como do tipo Psíquico, Dragão, Sombrio, Normal ou Fada.
A questão que desejamos responder é: um único atributo geral de um Pokémon seria parâmetro para se estabelecer um time de competição?

# Sobre o Dataset desse trabalho

O Dataset desse relatório leva em consideração os Pokémon da primeira até a sexta geração, totalizando 721 Pokémon, dos quais escolheremos 6. Realizando a contagem por combinação numérica, teremos um total de 191.082.439.565.304 times diferentes.

Os dados foram obtidos através dos jogos da franquia, os quais estão disponíveis nos seguintes websites:

* pokemon.com
* pokemondb
* bulbapedia

As informações contidas no Dataset sobre os Pokémon são as seguintes:

* '#': ID para cada Pokémon;
* Name: Nome para cada Pokémon;
* Type 1: Cada Pokémon possui um tipo, isso determina suas forças/fraquezas contra outros tipos;
* Type 2: Alguns Pokémon possuem tipo dual;
* Total: Soma de todos os status vem depois disso, um guia geral de quão forte um Pokémon é;
* HP: hit points, ou a vida de um Pokémon, define quanto dano um Pokémon suporta antes de desmaiar
* Attack: Atributo básico que define a força de ataques físicos;
* Defense: Atributo básico de resistência a ataques físicos;
* SP Atk: Ataque Especial, atributo básico de ataques elementais e não elementais;
* SP Def: Defesa Especial, atributo básico de resistência a ataques especiais;
* Speed: Velocidade, determina qual pokémon ataca primeiro.
---

# 1. Importando pacotes, importando dados e conhecendo a base de dados

In [ ]:
# Importando pacotes
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
# Importando base de dados
pokemon = pd.read_csv("https://raw.githubusercontent.com/askot19/Estudos/main/Pokemon.csv")
# Vendo os primeiros dados do Data Frame
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [ ]:
# Vendo os últimos dados do Data Frame, note que há lendários no Data Frame
pokemon.tail()

#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True

---

# 2. Selecionando os critérios estabelecidos

O primeiro critério de seleção é que o time seja composto por Pokémon não lendários, visto que estes são proibidos em competições oficiais. Desta maneira, vamos utilizar o pacote pandas para realizar esta seleção:

In [ ]:
# Selecionando Pokemon não lendários - NL
pokemon_nls = (pokemon['Legendary'] == False)

In [ ]:
# Verificando se há lendários no fim da lista, verificamos que os Pokémon lendários foram removidos de nossa seleção como esperado.
pokemon[pokemon_nls].tail()

#                 Name  Type 1  Type 2  Total  HP  Attack  Defense  \
787  711  GourgeistSuper Size   Ghost   Grass    494  85     100      122   
788  712             Bergmite     Ice     NaN    304  55      69       85   
789  713              Avalugg     Ice     NaN    514  95     117      184   
790  714               Noibat  Flying  Dragon    245  40      30       35   
791  715              Noivern  Flying  Dragon    535  85      70       80   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
787       58       75     54           6      False  
788       32       35     28           6      False  
789       44       46     28           6      False  
790       45       40     55           6      False  
791       97       80    123           6      False

In [ ]:
# Incluindo a seleção em um novo Data Frame, para mais fácil acesso e evitar exclusões não desejadas no nosso Data Frame Original
pokemon_nl = pokemon[pokemon_nls]

In [ ]:
# Verificando se o Data Frame está de acordo com o esperado
pokemon_nl.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Como fizemos a seleção, não se torna mais necessário repeti-la para novas análises. Podemos seguir para o próximo critério de seleção que é o Pokémon de maior ataque, o ideal é que este possua um relacionado a atributos físicos como lutador, pedra, metal, solo ou dragão. Mas o tipo, contudo, não é um critério excludente.

In [ ]:
# Verificando qual o maior valor de ataque da lista
pokemon_nl['Attack'].max()

185

In [ ]:
# Selecionando o Pokémon selecionado em um novo Data Frame
pokemon_mat = pokemon_nl[pokemon_nl['Attack'] == 185]

In [ ]:
pokemon_mat

#                     Name Type 1    Type 2  Total  HP  Attack  \
232  214  HeracrossMega Heracross    Bug  Fighting    600  80     185   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
232      115       40      105     75           2      False

Observamos que o Pokémon selecionado é o Mega Heracross dos tipos inseto e lutador, trazendo um dos melhores tipos associados ao atributo ataque.

<img alt = "mega heracross" width = "10%" src = "https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/abae6e28-7eb5-4262-8145-4dc4ac179c1d/d9ofs7x-fbe2e310-9d72-4791-9efa-0f09a0b75660.gif?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcL2FiYWU2ZTI4LTdlYjUtNDI2Mi04MTQ1LTRkYzRhYzE3OWMxZFwvZDlvZnM3eC1mYmUyZTMxMC05ZDcyLTQ3OTEtOWVmYS0wZjA5YTBiNzU2NjAuZ2lmIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.-iW_PcjC8tnNnBmnHL_YcX0_6rJ9QUJyXopN5-2Pbv0" align = "center">


Desta maneira podemos seguir para o próximo atributo que é a defesa, no intuito de obtermos o nosso melhor tanque, para tal o ideal é que o Pokémon escolhido seja dentre os tipos rocha, solo ou metal. O código utilizado é como se segue:

In [ ]:
# Verificando qual o valor da maior defesa
pokemon_nl['Defense'].max()

230

In [ ]:
# Selecionando os Pokémon de maior defesa para um Data Frame isolado
pokemon_mdef = pokemon_nl[pokemon_nl['Defense'] == 230]
pokemon_mdef

#                 Name Type 1  Type 2  Total  HP  Attack  Defense  \
224  208  SteelixMega Steelix  Steel  Ground    610  75     125      230   
230  213              Shuckle    Bug    Rock    505  20      10      230   
333  306    AggronMega Aggron  Steel     NaN    630  70     140      230   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
224       55       95     30           2      False  
230       10      230      5           2      False  
333       60       80     50           3      False

Das opções apresentadas, o que melhor se adequa é Mega Steelix, pois ele associa tanto o tipo metal quanto o tipo solo, combinando duas tipagens tanques e minimizando a sua fraqueza contra o tipo fogo.

<img alt = "mega steelix" width = "10%" src = "https://thumbs.gfycat.com/NiftyDigitalCalf-size_restricted.gif" align = "center">

In [ ]:
# Alterando a seleção para apenas um Pokémon de maior defesa
pokemon_mdef = pokemon_nl[pokemon_nl['Name'] == 'SteelixMega Steelix']
pokemon_mdef

#                 Name Type 1  Type 2  Total  HP  Attack  Defense  \
224  208  SteelixMega Steelix  Steel  Ground    610  75     125      230   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
224       55       95     30           2      False

Agora podemos dar continuidade em nossa busca para o Pokémon de maior ataque espcecial. O ideal é que este Pokémon seja não elemental, dentre os tipos psíquico, dragão, sombrio, fantasma ou então um elemental como água, fogo, grama, elétrico, gelo, solo, rocha ou mesmo dragão. No geral, os tipos não elementais são melhores adaptados para esta função, então focaremos neles.

In [ ]:
# Verificando qual o maior valor encontrado de Ataque Especial
pokemon_nl['Sp. Atk'].max()

175

In [ ]:
# Selecionando os Pokémon com maior Ataque Especial e verificando quais são
pokemon_mspa = pokemon_nl[pokemon_nl['Sp. Atk'] == 175]
pokemon_mspa

#                   Name   Type 1 Type 2  Total  HP  Attack  Defense  \
71  65  AlakazamMega Alakazam  Psychic    NaN    590  55      50       65   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
71      175       95    150           1      False

Temos como único membro não lendário com o maior Ataque Especial o Mega Alakazam, do tipo Psíquico. Uma excelente opção para essa role do time competitivo, manteremos esta opção.

<img alt = "mega alakazam" width = "20%" src = "https://thumbs.gfycat.com/WeeklyTangibleArgentineruddyduck-max-1mb.gif" align = "center">

Prosseguindo com o próximo, escolheremos aquele que possuir maior defesa especial, obtendo assim o Pokémon mais resistente.

In [ ]:
# Identificando qual a maior defesa especial
pokemon_nl['Sp. Def'].max()

230

In [ ]:
# Selecionando e verificando qual o Pokémon com o atributo defesa especial
pokemon_msdef = pokemon_nl[pokemon_nl['Sp. Def'] == 230]
pokemon_msdef

#     Name Type 1 Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
230  213  Shuckle    Bug   Rock    505  20      10      230       10      230   

     Speed  Generation  Legendary  
230      5           2      False

Das opções temos o Shuckle, um Pokémon dos tipos inseto e rocha. Devemos nos atentar, pois o Pokémon com maior ataque foi o Mega Heracross, que também possui o tipo inseto, desta forma não podemos repetir esta tipagem novamente.

<img alt = "shuckle" width = "10%" src = "https://static.wikia.nocookie.net/vees-total-pokemon/images/e/e2/Shuckle.gif/revision/latest/thumbnail/width/360/height/360?cb=20210627190233" align = "center">

Alguns pontos acerca do Shuckle é que ele possui atributos praticamente máximos para defesa e defesa especial, tornando-o um dos Pokémon mais resistentes da Dex. Entretanto, se observarmos os demais atributos é possível verificar que são valores absurdamente baixos, o que pode indicar que ele não seja a melhor opção. Contudo, manteremos ele na lista a fim de testar o nosso objetivo inicial.

Para o Pokémon estratégico, isto é, aquele com maior velocidade prosseguiremos com os mesmos procedimentos. Uma opção muito boa para Pokémon rápidos são os tipos voadores.

In [ ]:
# Verificando qual o maior valor encontrado de Velocidade
pokemon_nl['Speed'].max()

160

In [ ]:
# Selecionando e verificando qual é o Pokémon de maior Velocidade
pokemon_mvel = pokemon_nl[pokemon_nl['Speed'] == 160]
pokemon_mvel

#     Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
315  291  Ninjask    Bug  Flying    456  61      90       45       50   

     Sp. Def  Speed  Generation  Legendary  
315       50    160           3      False

Das opções temos o Ninjask, um Pokémon dos tipos inseto e voador. Devemos nos atentar, pois o tipo inseto não pode mais ser selecionado. Desta maneira, vamos selecionar o segundo maior valor, com intuito de substituí-lo.

In [ ]:
# Ordenando os 5 Pokémon mais rápidos e visualizando quais são eles.
pokemon_mvel = pokemon_nl.sort_values(by= ['Speed'], ascending = False).head(5)
pokemon_mvel

#                       Name   Type 1  Type 2  Total  HP  Attack  \
315  291                    Ninjask      Bug  Flying    456  61      90   
154  142  AerodactylMega Aerodactyl     Rock  Flying    615  80     135   
71    65      AlakazamMega Alakazam  Psychic     NaN    590  55      50   
19    15      BeedrillMega Beedrill      Bug  Poison    495  65     150   
678  617                   Accelgor      Bug     NaN    495  80      70   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
315       45       50       50    160           3      False  
154       85       70       95    150           1      False  
71        65      175       95    150           1      False  
19        40       15       80    145           1      False  
678       40      100       60    145           5      False

Como vimos a primeira opção Ninjask não é válida, pois trará novamente o tipo inseto para o time, tornando-o pouco diversificado. O segundo colocado Mega Aerodactyl é uma opção satisfatória para a competição, uma vez que ele é dos tipos voador e rocha.

<img alt = "mega aerodactyl" width = "20%" src = "https://64.media.tumblr.com/cec78d0e585802e732608c9d325006fa/tumblr_n0txgmYvGu1scncwdo1_500.gif" align = "center">

In [ ]:
# Selecionando o Mega Aerodactyl
pokemon_mvel = pokemon_nl[pokemon_nl['Name'] == 'AerodactylMega Aerodactyl']
pokemon_mvel

#                       Name Type 1  Type 2  Total  HP  Attack  \
154  142  AerodactylMega Aerodactyl   Rock  Flying    615  80     135   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
154       85       70       95    150           1      False

Como os times de campeonatos de combate Pokémon são realizados com equipes de 6 Pokémon, ainda precisamos selecionar um último componente. Para tal, vamos escolher um coringa, que apresente uma combinação de todos os melhores atributos possíveis na seguinte ordem Sp. Atk, Attack, Sp. Defense, Defense e Speed. O melhor candidato para um coringa seria um Pokémon do tipo Dragão, pois ele apresenta as maiores vantagens no quesito tipagem.

In [ ]:
# Selecionando e verificando quais Pokémon possuem primeiro ou segundo tipo iguais ao tipo Dragão
pokemon_sel1 = pokemon_nl[(pokemon_nl['Type 1'] == 'Dragon') | (pokemon_nl['Type 2'] == 'Dragon')]
pokemon_sel1

#                       Name    Type 1  Type 2  Total   HP  Attack  \
7      6  CharizardMega Charizard X      Fire  Dragon    634   78     130   
159  147                    Dratini    Dragon     NaN    300   41      64   
160  148                  Dragonair    Dragon     NaN    420   61      84   
161  149                  Dragonite    Dragon  Flying    600   91     134   
196  181      AmpharosMega Ampharos  Electric  Dragon    610   90      95   
249  230                    Kingdra     Water  Dragon    540   75      95   
275  254      SceptileMega Sceptile     Grass  Dragon    630   70     110   
360  329                    Vibrava    Ground  Dragon    340   50      70   
361  330                     Flygon    Ground  Dragon    520   80     100   
365  334                    Altaria    Dragon  Flying    490   75      70   
366  334        AltariaMega Altaria    Dragon   Fairy    590   75     110   
406  371                      Bagon    Dragon     NaN    300   45      75   
407  372                    Shelgon    Dragon     NaN    420   65      95   
408  373                  Salamence    Dragon  Flying    600   95     135   
409  373    SalamenceMega Salamence    Dragon  Flying    700   95     145   
491  443                      Gible    Dragon  Ground    300   58      70   
492  444                     Gabite    Dragon  Ground    410   68      90   
493  445                   Garchomp    Dragon  Ground    600  108     130   
494  445      GarchompMega Garchomp    Dragon  Ground    700  108     170   
671  610                       Axew    Dragon     NaN    320   46      87   
672  611                    Fraxure    Dragon     NaN    410   66     117   
673  612                    Haxorus    Dragon     NaN    540   76     147   
682  621                  Druddigon    Dragon     NaN    485   77     120   
694  633                      Deino      Dark  Dragon    300   52      65   
695  634                   Zweilous      Dark  Dragon    420   72      85   
696  635                  Hydreigon      Dark  Dragon    600   92     105   
761  691                   Dragalge    Poison  Dragon    494   65      75   
766  696                     Tyrunt      Rock  Dragon    362   58      89   
767  697                  Tyrantrum      Rock  Dragon    521   82     121   
774  704                      Goomy    Dragon     NaN    300   45      50   
775  705                    Sliggoo    Dragon     NaN    452   68      75   
776  706                     Goodra    Dragon     NaN    600   90     100   
790  714                     Noibat    Flying  Dragon    245   40      30   
791  715                    Noivern    Flying  Dragon    535   85      70   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
7        111      130       85    100           1      False  
159       45       50       50     50           1      False  
160       65       70       70     70           1      False  
161       95      100      100     80           1      False  
196      105      165      110     45           2      False  
249       95       95       95     85           2      False  
275       75      145       85    145           3      False  
360       50       50       50     70           3      False  
361       80       80       80    100           3      False  
365       90       70      105     80           3      False  
366      110      110      105     80           3      False  
406       60       40       30     50           3      False  
407      100       60       50     50           3      False  
408       80      110       80    100           3      False  
409      130      120       90    120           3      False  
491       45       40       45     42           4      False  
492       65       50       55     82           4      False  
493       95       80       85    102           4      False  
494      115      120       95     92           4      False  
671       60       30       40     57           5  

In [ ]:
# Selecionando e ordenando Pokémon do tipo dragão de acordo com a ordem prioridade de atributos escolhidos
pokemon_cor = pokemon_sel1.sort_values(by = ['Sp. Atk', 'Attack', 'Sp. Def', 'Defense', 'Speed'], ascending = False).head(5)
pokemon_cor

#                       Name    Type 1  Type 2  Total   HP  Attack  \
196  181      AmpharosMega Ampharos  Electric  Dragon    610   90      95   
275  254      SceptileMega Sceptile     Grass  Dragon    630   70     110   
7      6  CharizardMega Charizard X      Fire  Dragon    634   78     130   
696  635                  Hydreigon      Dark  Dragon    600   92     105   
494  445      GarchompMega Garchomp    Dragon  Ground    700  108     170   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
196      105      165      110     45           2      False  
275       75      145       85    145           3      False  
7        111      130       85    100           1      False  
696       90      125       90     98           5      False  
494      115      120       95     92           4      False

Dentre os possíveis casos, o que possui os melhores atributos pela ordem escolhida é o Mega Ampharos, dos tipos Elétrico e Dragão. Permaneceremos com ele no time a fim dele ser o nosso Pokémon coringa e possuir tipagens diversificadas do restante do time.

<img alt = "mega aerodactyl" width = "9%" src = "https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/f19fb800-4e27-4488-9f51-edf5f95d6d54/d79s3ry-66086952-5848-44be-bbbb-62829b96aadf.gif?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcL2YxOWZiODAwLTRlMjctNDQ4OC05ZjUxLWVkZjVmOTVkNmQ1NFwvZDc5czNyeS02NjA4Njk1Mi01ODQ4LTQ0YmUtYmJiYi02MjgyOWI5NmFhZGYuZ2lmIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.fanAfp_K2506edyq4KAZkzeP0Aj_TwFZVsz7Eh6GNcE" align = "center">

In [ ]:
pokemon_cor = pokemon_sel1[pokemon_sel1['Name'] == 'AmpharosMega Ampharos']
pokemon_cor

#                   Name    Type 1  Type 2  Total  HP  Attack  Defense  \
196  181  AmpharosMega Ampharos  Electric  Dragon    610  90      95      105   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
196      165      110     45           2      False

In [ ]:
# Concatenando Data Frame para apresentar um Time Único
pokemon_team = pd.concat([pokemon_mat, pokemon_mdef, pokemon_msdef, pokemon_mspa, pokemon_mvel, pokemon_cor], ignore_index = True)
pokemon_team

#                       Name    Type 1    Type 2  Total  HP  Attack  \
0  214    HeracrossMega Heracross       Bug  Fighting    600  80     185   
1  208        SteelixMega Steelix     Steel    Ground    610  75     125   
2  213                    Shuckle       Bug      Rock    505  20      10   
3   65      AlakazamMega Alakazam   Psychic       NaN    590  55      50   
4  142  AerodactylMega Aerodactyl      Rock    Flying    615  80     135   
5  181      AmpharosMega Ampharos  Electric    Dragon    610  90      95   

   Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0      115       40      105     75           2      False  
1      230       55       95     30           2      False  
2      230       10      230      5           2      False  
3       65      175       95    150           1      False  
4       85       70       95    150           1      False  
5      105      165      110     45           2      False

O time selecionado é então composto por Mega Heracross, Mega Steelix, Shuckle, Mega Alakazam, Mega Aerodactyl e Mega Ampharos.

<img alt = "mega heracross" width = "10%" src = "https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/abae6e28-7eb5-4262-8145-4dc4ac179c1d/d9ofs7x-fbe2e310-9d72-4791-9efa-0f09a0b75660.gif?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcL2FiYWU2ZTI4LTdlYjUtNDI2Mi04MTQ1LTRkYzRhYzE3OWMxZFwvZDlvZnM3eC1mYmUyZTMxMC05ZDcyLTQ3OTEtOWVmYS0wZjA5YTBiNzU2NjAuZ2lmIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.-iW_PcjC8tnNnBmnHL_YcX0_6rJ9QUJyXopN5-2Pbv0" align = "center"> <img alt = "mega steelix" width = "10%" src = "https://thumbs.gfycat.com/NiftyDigitalCalf-size_restricted.gif" align = "center"> <img alt = "mega alakazam" width = "15%" src = "https://thumbs.gfycat.com/WeeklyTangibleArgentineruddyduck-max-1mb.gif" align = "center"> <img alt = "shuckle" width = "10%" src = "https://static.wikia.nocookie.net/vees-total-pokemon/images/e/e2/Shuckle.gif/revision/latest/thumbnail/width/360/height/360?cb=20210627190233" align = "center"> <img alt = "mega aerodactyl" width = "20%" src = "https://64.media.tumblr.com/cec78d0e585802e732608c9d325006fa/tumblr_n0txgmYvGu1scncwdo1_500.gif" align = "center"> <img alt = "mega aerodactyl" width = "9%" src = "https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/f19fb800-4e27-4488-9f51-edf5f95d6d54/d79s3ry-66086952-5848-44be-bbbb-62829b96aadf.gif?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcL2YxOWZiODAwLTRlMjctNDQ4OC05ZjUxLWVkZjVmOTVkNmQ1NFwvZDc5czNyeS02NjA4Njk1Mi01ODQ4LTQ0YmUtYmJiYi02MjgyOWI5NmFhZGYuZ2lmIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.fanAfp_K2506edyq4KAZkzeP0Aj_TwFZVsz7Eh6GNcE" align = "center">

---
# 3. Comparando o time selecionado com o META

Com a definição do nosso time, vamos agora determinar se eles se encaixam nas competições. Para isso, utilizaremos a plataforma [pvpoke](https://pvpoke.com/rankings/all/2500/overall/), na categoria Ultraleague que compete aos Pokémon da sexta geração.
Basicamente, os Pokémon são rankeados com um score que varia de 0 a 100 e são classificados de acordo com esta pontuação e considerados com o melhor conjunto de moveset e habilidades para se extrair o maior potencial possível de cada. A classificação obtida é a seguinte:

1. Mega Heracross - Posição #89 - Score 84,9
2. Mega Steelix - Posição #11 - Score 92,6
3. Mega Alakazam - Posição #396 - Score 70,3
4. Shuckle - NÃO CLASSIFICADO - SEM SCORE
5. Mega Aerodactyl - Posição #456 - Score 66,6
6. Mega Ampharos - Posição #179 - Score 80,3

Como podemos ver, em sua maioria, os Pokémon obtidos através da maximização dos atributos únicos pertencem ao META, excetuando-se o Shuckle. Estes fatores são indicativos que uma seleção, possivelmente deve ser feita por meio dos atributos individuais dos Pokémon.

---

# 4. Resultados e conclusões

Realizamos uma simulação de batalha da equipe com o intuito de se obter uma estatística das possibilidades de vitória. A simulação poderia ser realizada com todos os 721 Pokémon do banco de dados, contudo, para esta análise consideraremos apenas o embate contra os 20 primeiros Pokémon do META.

Estes critérios nos fornecerão um total de 120 batalhas, deste total de batalhas consideraremos o número de vitórias para extrair o percentual de vitórias do time. As simulações foram feitas através da Ultra League do site [pvpoke](https://pvpoke.com/rankings/all/2500/overall/).

In [ ]:
# Carregando os resultados das simulações das 120 batalhas Pokémon

vitorias = { 'Alakazam': {
             0: 'd',
             1: 'd',
             2: 'd',
             3: 'd',
             4: 'd',
             5: 'd',
             6: 'd',
             7: 'd',
             8: 'd',
             9: 'd',
             10: 'd',
             11: 'd',
             12: 'd',
             13: 'd',
             14: 'd',
             15: 'd',
             16: 'v',
             17: 'v',
             18: 'v',
             19: 'v'},
             'Aerodactyl': {
                          0: 'v',
                          1: 'v',
                          2: 'v',
                          3: 'd',
                          4: 'd',
                          5: 'd',
                          6: 'd',
                          7: 'd',
                          8: 'd',
                          9: 'd',
                          10: 'd',
                          11: 'd',
                          12: 'd',
                          13: 'd',
                          14: 'd',
                          15: 'd',
                          16: 'd',
                          17: 'd',
                          18: 'd',
                          19: 'd'},
             'Ampharos': {
                        0: 'v',
                        1: 'v',
                        2: 'v',
                        3: 'v',
                        4: 'v',
                        5: 'd',
                        6: 'd',
                        7: 'd',
                        8: 'd',
                        9: 'd',
                        10: 'd',
                        11: 'd',
                        12: 'd',
                        13: 'd',
                        14: 'd',
                        15: 'd',
                        16: 'd',
                        17: 'd',
                        18: 'd',
                        19: 'd'},
             'Heracross': {
                         0: 'd',
                         1: 'd',
                         2: 'd',
                         3: 'd',
                         4: 'd',
                         5: 'd',
                         6: 'd',
                         7: 'd',
                         8: 'd',
                         9: 'v',
                         10: 'v',
                         11: 'v',
                         12: 'v',
                         13: 'v',
                         14: 'v',
                         15: 'v',
                         16: 'v',
                         17: 'v',
                         18: 'v',
                         19: 'v'},
             'Steelix': {
                       0: 'v',
                       1: 'v',
                       2: 'v',
                       3: 'v',
                       4: 'v',
                       5: 'v',
                       6: 'd',
                       7: 'd',
                       8: 'd',
                       9: 'd',
                       10: 'd',
                       11: 'd',
                       12: 'd',
                       13: 'd',
                       14: 'd',
                       15: 'd',
                       16: 'd',
                       17: 'd',
                       18: 'd',
                       19: 'd'},
             'Shuckle': {
                       0: 'd',
                       1: 'd',
                       2: 'd',
                       3: 'd',
                       4: 'd',
                       5: 'd',
                       6: 'd',
                       7: 'd',
                       8: 'd',
                       9: 'd',
                       10: 'd',
                       11: 'd',
                       12: 'd',
                       13: 'd',
                       14: 'd',
                       15: 'd',
                       16: 'd',
                       17: 'd',
                       18: 'd',
                       19: 'd'}
             }

In [ ]:
# Importando resultado para um novo Data Frame
batalhas = pd.DataFrame(vitorias)
batalhas

Alakazam Aerodactyl Ampharos Heracross Steelix Shuckle
0         d          v        v         d       v       d
1         d          v        v         d       v       d
2         d          v        v         d       v       d
3         d          d        v         d       v       d
4         d          d        v         d       v       d
5         d          d        d         d       v       d
6         d          d        d         d       d       d
7         d          d        d         d       d       d
8         d          d        d         d       d       d
9         d          d        d         v       d       d
10        d          d        d         v       d       d
11        d          d        d         v       d       d
12        d          d        d         v       d       d
13        d          d        d         v       d       d
14        d          d        d         v       d       d
15        d          d        d         v       d       d
16        v          d        d         v       d       d
17        v          d        d         v       d       d
18        v          d        d         v       d       d
19        v          d        d         v       d       d

In [ ]:
# Convertendo vitórias e derrotas para booleanos
batalhas.replace('d', 0, inplace = True)
batalhas.replace('v', 1, inplace = True)
batalhas

Alakazam  Aerodactyl  Ampharos  Heracross  Steelix  Shuckle
0          0           1         1          0        1        0
1          0           1         1          0        1        0
2          0           1         1          0        1        0
3          0           0         1          0        1        0
4          0           0         1          0        1        0
5          0           0         0          0        1        0
6          0           0         0          0        0        0
7          0           0         0          0        0        0
8          0           0         0          0        0        0
9          0           0         0          1        0        0
10         0           0         0          1        0        0
11         0           0         0          1        0        0
12         0           0         0          1        0        0
13         0           0         0          1        0        0
14         0           0         0          1        0        0
15         0           0         0          1        0        0
16         1           0         0          1        0        0
17         1           0         0          1        0        0
18         1           0         0          1        0        0
19         1           0         0          1        0        0

In [ ]:
# Calculando o percentual de vitórias de cada Pokémon
percentual_vitorias = batalhas.sum()/20*100
percentual_vitorias

Alakazam      20.0
Aerodactyl    15.0
Ampharos      25.0
Heracross     55.0
Steelix       30.0
Shuckle        0.0
dtype: float64

## 5. Considerações finais

Podemos observar que as taxas de vitórias não são satisfatórias para cada Pokémon. Alakazam apresenta 20%, Aerodactyl 15%, Ampharos 25%, Heracross 55%, Steelix 30% e Shuckle 0%.

Com exceção do Mega Heracross, todos os Pokémon apresentam taxa de sucesso inferior a 50%, embora eles possuam os melhores atributos únicos possíveis. Isto acontece, pois não apenas os atributos importam para o sucesso de uma batalha, mas também as habilidades que este Pokémon sustenta, os ataques, sua tipagem e também o fator sorte.

Atualmente a melhor forma de se definir um bom time é utilizando diretamente a listagem estabelecida pelos sites, e mesmo assim existe a possibilidade de não se ter uma taxa de sucesso muito satisfatória.

Uma implementação futura para este projeto seria a aplicação de Machine Learning, a fim de se considerar todos os fatores que influenciam as taxas de vitória, construindo o melhor time possível para um campeonato.

Ademais, concluímos que a utilização de um atributo único não se apresenta como um bom fator determinante para equipes, sendo a maior evidência a escolha do Pokémon Shuckle que não apresentou qualquer valor de contribuição para a equipe Pokémon, muito provável que tenha sido a pior escolha possível dentre as 721 opções.